### ISSUE TITLE
Next level of scale: run the `baseline` test plan on a 10k node network (cluster:k8s runner)

### ISSUE THREAD

**Issue by raulk** [ipfs/testground#599]
Dependent on #597.

**Comment by raulk** [comment]
Some of the areas of trouble / hotspots to monitor while we run the experiments, along with some ideas.

Preamble: the experiment consists in running all test cases of the baseline tests, on the cluster:k8s runner, with instance count 10k (probably getting there gradually; we've already tested 5k with the "network initialized" tests).

* ✅ Redis: Redis config and system config have been adjusted in `cluster:k8s` to deal with the issues we detected in our scaling experiment: https://github.com/ipfs/testground/issues/727. We scaled up to 80k connections in my local docker instance. I expect Redis will be no source of setbacks.

* ❗️ Networks (flannel, weave, CNIgenie): I think this is a risky area, both at the management level (API calls to attach and detach containers -- are these components able to deal with a sudden influx of 10k calls in a short time span?).
  * Enable the highest level of logging verbosity at the flannel, weave and CNIgenie levels. Keep a close eye on logs.
  * There are two interactions with the management plane of these networks: one by k8s when the pod is starting, one by sidecar. Keep a close eye on all of them.

* ❗️ Resource quotas: this is the area I'm least confident about right now. We do not have any tools to grok the resource usage of our test plans, therefore we're flying blindly. The k8s bin-packing algorithm relies on the resource quotas we set, which folks are not even adjusting in their compositions.
  * Potential feature: add a new output asset `sys_diag.out` (system diagnostics), that records a JSON-based system stats dump of the container (CPU usage, memory usage, network usage), under the Docker-based runners.
  * Library: https://github.com/shirou/gopsutil for the Go SDK.
  * Create a `testground analyse-resource-usage` command that takes the collected tgz/zip and extracts insights.
  * We absolutely need this tooling.
    * Setting resource quotas on test plans should be on the test plan writers. Testground cannot predict the resource consumption of a given test plan.
    * We can expect them to grok at Prometheus consoles, and make sense of what they see. That's too complex. There's no reason why this data needs to be processes graphically. We need a turnkey approach here, and Testground oughts to provide the tooling to assess test case footprint.

* ⚠️ Sidecar: I'm less concerned with the operation of the sidecar itself, because it's a per-node component, so each sidecar will only have to deal with less than 200 nodes at a time.
  * However, worth keeping an eye on the logs.
  * Problems in sidecar will manifest as tests not progressing to the "network initialized" stage.

* ⚠️ Kubernetes itself: I don't know if k8s can cope with instantly scheduling an influx of 10k pods, where the node fleet will need to scale. It can certainly cope with the management of such workloads, but I'm less confident about the aggressive scheduling we're subjecting it to.

* ⚠️ AWS: will the Autoscaling group cope properly? Keep an eye on the AWS console. Probably nothing to worry about here, but it's worth watching what happens, how long the upscaling takes, and documenting our observations.

* ⚠️ Pushgateway: we only run a single pushgateway instance, which will suddenly see a spike of 10k processes pushing data to it. It might fall over.
  * Keep an eye on the pod logs.

**Comment by nonsense** [comment]
A few notes:
1. AWS auto-scaling - we don't have any auto-scaling right now. This is done manually.
2. Kubernetes - we limit the number of API request to 32 concurrent requests, which is just fine. We haven't seen any errors reported from the API so far and haven't had any issues scheduling workloads. https://kubernetes.io/docs/setup/best-practices/cluster-large/ - we should be just fine with 10k pods and beyond.
3. Redis - the configs you ran on your local machine are different to what we have in https://github.com/ipfs/testground/blob/master/infra/k8s/testground-infra/values.yaml , so there is definitely room for errors and misconfigurations. Just the other day we forgot to set `usePassword` to `false`, which was failing all the `sidecars`. Testing the sync service in a distributed environment vs locally with the out-of-the-box `redis-benchmark` is a rather different use-case - we never benchmarked XREAD operations, which is one of the primary operations we are using from the sync service.
4. Resource quotas - agree that we don't have good intuition on what are good values here.
5. Networks - agree that we need better visibility there. For now we rely on logs and the Weave dashboard.

**Comment by nonsense** [comment]
@raulk this is a good example of the number of moving pieces in Testground - this is why it is *crucial* to establish good foundations for infrastructure metrics (including dashboards), and testplan metrics (including dashboards), and logs aggregation (needed to detect unknown errors).

We have done a good job so far, but these tasks are far from complete.

At the moment, when I run a single 10k benchmark it is hard to even say if it succeeded or not - we don't have tools to say if all testplan instances returned `outcome ok`, so we write custom shell scripts (every developer for themselves). If we had a dashboard - it would be a simple count that you can confirm, that would work out of the box for example. Unless you notice one of the tens of log files having emitted an ERROR message, you might miss a failure.

---

I think it is a good exercise to see what fails when we try to run 10k benchmarks, as things ultimately start to fail for various reasons (such as resource requirements for 1 of the 10 subsystems ; incorrect size of cluster ; low values for resource requirements of testplans, which requires scaling ; incorrectly set timeouts, etc. etc.), but we have to establish good baselines for lower scales (1k, 2k, 5k) with proper tooling -
1. infrastructure metrics with dashboards
2. application (benchmark testplan, dht testplan) metrics with dashboards
3. logs aggregation and search for errors

**Comment by raulk** [comment]
@nonsense I don't think anything is holding us back? Mind summarising your thoughts into a concrete action plan?

**Comment by nonsense** [comment]
My thoughts overall are that this ticket implicitly suggests that we should not just run the benchmarks testplans, but also collect data (success/failure), analyse the outputs, and establish some baselines.

Getting this signal out of the benchmark testplan runs right now is very difficult, because our tooling is not there yet, and we haven't established baselines for even lower scales.

Each benchmarks/testplan run (single benchmark of a single testcase) takes 10-20min. so this is extremely time consuming, before having optimised the extraction of signal from these runs. This is my concern.

**Comment by nonsense** [comment]
TLDR:

10k testplans of the `storm` testplan are passing, as long as we don't have broken `weave` instances and `drain` the nodes where `weave` fails

Next steps:
- fix `weave` bugs or account for them on `testground` level (detect and drain nodes?)
- fix `infrastructure monitoring` on larger clusters
- try smaller instances (instead of c5.2xlarge - maybe c5.xlarge if there are enough available)
- add logs aggregation (filebeat + ES)

---

I've ran a lot of tests over the last 2 weeks, after we introduced the new sync service and increased many of the TCP/UDP stack limits (buffers, queues, etc.)

A few learnings from this exercise:
1. When running clusters of ~150 VMs and above, it helps to use c5.4xlarge for master, in order to have faster scheduling of pods (https://kubernetes.io/docs/setup/best-practices/cluster-large/)
2. When running clusters of >150 VMs sometimes we hit issues with:
- AWS (not provisioning enough VMs) - this is not visible to the Testground operator, unless they monitor AWS
- Weave sometimes enters in CrashLoopBackoff state (due to taken port on host VM) - something we have to investigate
- Weave sometimes deadlocks and never recovers from the 0/1 state (i.e. that /status endpoint is dead)
    - Workaround is - watch for pods in 0/1 state and drain the node if you see Weave
- Sidecar sometimes fails (possibly due to weave? possibly due to the number of barriers it has to follow?)
3. Using the sync service to publish 10k node information structs, and then get 10k*10k structs out of it, is somewhat slow.
4. Prometheus goes OOM almost immediately (after 1-2 runs), without me having opened even a single Grafana dashboard that would query it, which rules out query problems for now (we will probably have these issues as well in the future)
prometheus-operator-kube-state-metrics-7f979567df-kbjgl              0/1     CrashLoopBackOff   8          101m    100.96.45.3      ip-172-20-53-83.eu-west-2.compute.internal    <none>           <none>
prometheus-prometheus-operator-prometheus-0                          2/3     CrashLoopBackOff   15         101m    100.96.100.5     ip-172-20-39-207.eu-west-2.compute.internal   <none>           <none>

---

Here are a few of the results from the `storm` testplan:

---

120 VMs worker nodes - c5.2xlarge

Pod resources:
memory = '140Mi'
cpu = '70m'

5000 testplan instances
Listening on 10 sockets each ; Publishing one struct with node information to sync service ;
Making 50 outgoing connections
Sending/Reading 5/10/50kb across each connection

---

7000 testplan instances
Listening on 10 sockets each ; Publishing one struct with node information to sync service ;
Making 50 outgoing connections
Sending/Reading 2kb across each connection

---

8000 testplan instances -> FAIL on barrier (I somehow killed Weave at this point)

---

180 VMs worker nodes - c5.2xlarge

Pod resources:
memory = '140Mi'
cpu = '70m'

7000 testplan instances
Listening on 10 sockets each ; Publishing one struct with node information to sync service ;
Making 50 outgoing connections
Sending/Reading 2kb across each connection

---

230 VMs worker nodes - c5.2xlarge

Pod resources:
memory = '150Mi'
cpu = '100m'

44 pods per VM

10000 testplan instances
Listening on 10 sockets each ; Publishing one struct with node information to sync service ;
Making 50 outgoing connections
Sending/Reading 2/10kb across each connection

[screenshots-10k.zip](https://github.com/testground/testground/files/4546153/screenshots-10k.zip)

**Comment by Robmat05** [comment]
Sliding this to v0.6 and adding it as top priority for @nonsense to add the benchmarks + storm test plan merged with benchmarks plan as discussed in our Launch Sync yesterday
